### PR TITLE
Feature/OAuth2

### PR DESCRIPTION
**Pull request by fgaudreault**
- Ready for review
- Missing documentation

### CODE REVIEW

**Comment by fgaudreault** [comment]
Doc is there, Package is there.  Please review.

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 1]
```diff
@@ -1960,6 +1960,51 @@ Here are the different configuration parameters that can be set for each portal
 
 CAUTION: If you set a custom template_path, make sure to copy over all of the default template files in the new custom folder. PacketFence will only render templates from the files in that custom folder if the value is setted. If files are missing, you'll be facing some error in templates generation. 
 
+OAuth2 Authentication
+~~~~~~~~~~~~~~~~~~~~~
+
+Some public captive portal provides the capability of allowing a guest/user to register using his Google or Facebook account.  We can also do that in PacketFence.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: "The captive portal of PacketFence allows a guest/user to register using his Google or Facebook account."

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 2]
```diff
@@ -1960,6 +1960,51 @@ Here are the different configuration parameters that can be set for each portal
...
+   client_secret = YOURAPISECRET
+   site = https://accounts.google.com
+   authorize_path = /o/oauth2/auth
+   access_token_path = /o/oauth2/token
+   scope = https://www.googleapis.com/auth/userinfo.email
+   protected_resource_url = https://www.googleapis.com/oauth2/v2/userinfo
+   access_token_param = oauth_token
+   redirect_uri = https://o2.packetfence.org/o2/google
+
+Also, enable the Google authentication in the guests_self_registration.modes field.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: .. in the `guests_self_registration.modes` field.

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 3]
```diff
@@ -1960,6 +1960,51 @@ Here are the different configuration parameters that can be set for each portal
...
+   [oauth2 facebook]
+   client_id = YOURAPPID
+   client_secret = YOURAPPSECRET
+   site = https://graph.facebook.com
+   access_token_path = /oauth/access_token
+   scope = email
+   access_token_param = access_token
+   redirect_uri = https://o2.packetfence.org/o2/facebook
+
+Also, enable the Facebook authentication in the guests_self_registration.modes field.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: .. in the `guests_self_registration.modes` field.

File: lib/pf/web.pm [hunk 4]
```diff
@@ -391,6 +393,63 @@ sub generate_admin_error_page {
...
+        my $json_text = $json->decode($response->content());
+        if ($provider eq 'google') {
+            $logger->info("OAuth2 successfull, register and release for email $json_text->{email}");
+            return ($TRUE,$json_text->{email});
+        } elsif ($provider eq 'facebook') {
+            $logger->info("OAuth2 successfull, register and release for username $json_text->{username}");
+            return ($TRUE,$json_text->{username});
+        }
+   } else {
+        $logger->info("OAuth2 failed, redireting to login page with appropriate error");
```
Comment (fgaudreault) [inline lib/pf/web.pm]: Woops, I need to fix that :P

File: html/captive-portal/oauth2.cgi [hunk 5]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+OAuth2.cgi
```
Comment (cgx) [inline html/captive-portal/oauth2.cgi]: The cgi script name is actually in lower case.

File: lib/pf/web.pm [hunk 6]
```diff
@@ -391,6 +393,63 @@ sub generate_admin_error_page {
...
+   #Get the token
+   my $token = oauth2_client($provider)->get_access_token($portalSession->getCgi()->url_param('code'));
+
+   my $response;
+
+   # Validate the token
+   if ($provider eq 'google') {
+       $response = $token->get('https://www.googleapis.com/oauth2/v2/userinfo');
+   } elsif ($provider eq 'facebook') {
+       $response = $token->get('https://graph.facebook.com/me');
```
Comment (cgx) [inline lib/pf/web.pm]: Shall the two providers URLs be configurable from pf.conf?

**Comment by fgaudreault** [comment]
Remaining task :
- Test with Inline into play

**Comment by fgaudreault** [comment]
I think we are good to go now.  Please do an extensive and final review to merge.

**Comment by dwlfrth** [comment]
As a reminder "do not forget style": adapt with the new mod_perl stuff

**Comment by dwlfrth** [comment]
do you mind if we rename all of the names (oauth, o2, auth2...) to oauth2 for consistency in the code?

**Comment by fgaudreault** [comment]
You can :)

**Pull request closed without merging**